### ISSUE TITLE
Unexpected allocations in v0.6 with sum(::Function, ::Vector) and maximum(::Function, ::Vector)

### ISSUE THREAD

**Issue by rdeits** [JuliaLang/julia#21170]
I've found a small performance regression that happens to be slowing down the inner loop of https://github.com/ahwillia/NonNegLeastSquares.jl/pull/3 in v0.6. The MRE is:

on Julia v0.5:

```julia
julia> using BenchmarkTools

julia> x = rand(5);

julia> @benchmark maximum(abs, $x)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     9.971 ns (0.00% GC)
  median time:      10.012 ns (0.00% GC)
  mean time:        10.699 ns (0.00% GC)
  maximum time:     40.841 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     999
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

with Julia v0.6:

```julia
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.6.0-pre.alpha.260 (2017-03-25 22:14 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit fb312c40aa (1 day old master)
|__/                   |  x86_64-apple-darwin13.4.0

julia> versioninfo()
Julia Version 0.6.0-pre.alpha.260
Commit fb312c40aa (2017-03-25 22:14 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin13.4.0)
  CPU: Intel(R) Core(TM) i7-2860QM CPU @ 2.50GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Sandybridge)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, sandybridge)

julia> x = rand(5);

julia> using BenchmarkTools

julia> @benchmark maximum(abs, $x)
BenchmarkTools.Trial:
  memory estimate:  16 bytes
  allocs estimate:  1
  --------------
  minimum time:     62.830 ns (0.00% GC)
  median time:      64.047 ns (0.00% GC)
  mean time:        67.428 ns (1.37% GC)
  maximum time:     1.533 μs (91.28% GC)
  --------------
  samples:          10000
  evals/sample:     979
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

Those allocations also show up when running with `--track-allocation=user` and when running similar code in a real function (which otherwise would not allocate at all). 

I see similar allocations with `sum(::Function)`. 

with Julia v0.5:

```julia
julia> @benchmark sum(abs, $x)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     6.141 ns (0.00% GC)
  median time:      6.143 ns (0.00% GC)
  mean time:        6.222 ns (0.00% GC)
  maximum time:     14.444 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1000
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

with Julia v0.6:

```julia
julia> @benchmark sum(abs, $x)
BenchmarkTools.Trial:
  memory estimate:  16 bytes
  allocs estimate:  1
  --------------
  minimum time:     639.036 ns (0.00% GC)
  median time:      645.725 ns (0.00% GC)
  mean time:        666.803 ns (0.00% GC)
  maximum time:     2.387 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     167
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

**Comment by martinholters** [comment]
Has this fixed been by #21167?

**Comment by rdeits** [comment]
Indeed it is! Looks like my v0.6 was just a few hours too old. Thanks!
